subsys/emds/emds.c

```diff
@@ -11,60 +11,66 @@
 #include <drivers/flash.h>
 #include "emds_flash.h"
 
-#if defined(CONFIG_BT) && !defined(CONFIG_BT_LL_SW_SPLIT)
-#include <sdc.h>
-#include <mpsl.h>
-#endif
-
 #include <logging/log.h>
 LOG_MODULE_REGISTER(emds, CONFIG_EMDS_LOG_LEVEL);
 
 K_SEM_DEFINE(emds_sem, 0, 1);
 static bool emds_ready;
 static bool emds_initialized;
-static bool emds_complete;
+static uint32_t store_key;
 
 static sys_slist_t emds_dynamic_entries;
 static struct emds_fs emds_flash;
+static emds_store_cb_t app_store_cb;
 
 static void emds_handler(void)
 {
-	k_sem_reset(&emds_sem);
-	k_sem_take(&emds_sem, K_FOREVER);
+	while (true) {
+		k_sem_reset(&emds_sem);
+		k_sem_take(&emds_sem, K_FOREVER);
 
-	k_sched_lock();
+		k_sched_lock();
 
-	LOG_DBG("Emergency Data Storeage released");
+		LOG_DBG("Emergency Data Storeage released");
 
-	STRUCT_SECTION_FOREACH(emds_entry, ch) {
-		ssize_t len = emds_flash_write(&emds_flash,
-					ch->id, ch->data, ch->len);
-		if (len < 0) {
-			LOG_ERR("Write static entry: (%d) error (%d)",
-				ch->id, len);
-		} else if (len != ch->len) {
-			LOG_ERR("Write static entry: (%d) failed (%d:%d)",
-				ch->id, ch->len, len);
+		STRUCT_SECTION_FOREACH(emds_entry, ch) {
+			ssize_t len = emds_flash_write(&emds_flash,
+						       ch->id, ch->data, ch->len);
+			if (len < 0) {
+				LOG_ERR("Write static entry: (%d) error (%d)",
+					ch->id, len);
+			} else if (len != ch->len) {
+				LOG_ERR("Write static entry: (%d) failed (%d:%d)",
+					ch->id, ch->len, len);
+			}
 		}
-	}
 
-	struct emds_dynamic_entry *ch;
+		struct emds_dynamic_entry *ch;
 
-	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
-		ssize_t len = emds_flash_write(&emds_flash,
-					ch->entry.id, ch->entry.data, ch->entry.len);
-		if (len < 0) {
-			LOG_ERR("Write dynamic entry: (%d) error (%d).",
-				ch->entry.id, len);
+		SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+			ssize_t len = emds_flash_write(&emds_flash,
+						       ch->entry.id, ch->entry.data, ch->entry.len);
+			if (len < 0) {
+				LOG_ERR("Write dynamic entry: (%d) error (%d).",
+					ch->entry.id, len);
+			}
+			if (len != ch->entry.len) {
+				LOG_ERR("Write dynamic entry: (%d) failed (%d:%d).",
+					ch->entry.id, ch->entry.len, len);
+			}
 		}
-		if (len != ch->entry.len) {
-			LOG_ERR("Write dynamic entry: (%d) failed (%d:%d).",
-				ch->entry.id, ch->entry.len, len);
+
+		emds_ready = false;
+
+		k_sched_unlock();
+
+		/* Unlock all interrupts */
+		irq_unlock(store_key);
+
+		if (app_store_cb) {
```

Comment:
I'm just curious will it be better to call callback before interrupts enabling? I guess after irqs enabling a lot of them will be triggered immediately.